eks/deploy_state.go

```diff
@@ -0,0 +1,323 @@
+package eks
+
+import (
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/elb"
+	"github.com/aws/aws-sdk-go/service/elbv2"
+	"github.com/gruntwork-io/go-commons/errors"
+	"github.com/gruntwork-io/kubergrunt/kubectl"
+	"github.com/gruntwork-io/kubergrunt/logging"
+	"github.com/sirupsen/logrus"
+	"io/ioutil"
+	"k8s.io/apimachinery/pkg/util/json"
+	"os"
+	"strings"
+	"time"
+)
+
+const defaultStateFile = "./.kubergrunt.state"
+
+type DeployState struct {
+	GatherASGInfoDone      bool
+	SetMaxCapacityDone     bool
+	ScaleUpDone            bool
+	WaitForNodesDone       bool
+	CordonNodesDone        bool
+	DrainNodesDone         bool
+	DetachInstancesDone    bool
+	TerminateInstancesDone bool
+	RestoreCapacityDone    bool
+
+	Path string
+	ASGs []ASG
+
+	maxRetries          int
+	sleepBetweenRetries time.Duration
+
+	logger *logrus.Entry
+}
+
+type ASG struct {
+	Name                 string
+	OriginalCapacity     int64
+	MaxCapacityForUpdate int64
+	OriginalMaxCapacity  int64
+	OriginalInstances    []string
+	NewInstances         []string
+}
+
+func initDeployState(file string, ignoreExistingFile bool, maxRetries int, sleepBetweenRetries time.Duration) (*DeployState, error) {
+	logger := logging.GetProjectLogger()
+	var deployState *DeployState
+
+	if ignoreExistingFile {
+		logger.Info("Ignore existing state file.")
+		deployState = newDeployState(file)
+	} else {
+		logger.Debugf("Looking for existing recovery file %s", file)
+		data, err := ioutil.ReadFile(file)
+		if err != nil {
+			logger.Debugf("No state present, creating new: %s", err.Error())
+			deployState = newDeployState(file)
+		} else {
+			var parsedState DeployState
+			err = json.Unmarshal(data, &parsedState)
+			if err != nil {
+				return nil, err
+			}
+			deployState = &parsedState
+		}
+	}
+
+	deployState.logger = logger
+	deployState.maxRetries = maxRetries
+	deployState.sleepBetweenRetries = sleepBetweenRetries
+
+	return deployState, nil
+}
+
+func (state *DeployState) persist() error {
+	file := state.Path
+	state.logger.Debugf("storing state file %s", file)
+
+	data, err := json.Marshal(state)
+
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	err = ioutil.WriteFile(file, data, 0644)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+	return nil
+}
+
+func (state *DeployState) delete() error {
+	file := state.Path
+	state.logger.Debugf("Deleting state file %s", file)
+
+	err := os.Remove(file)
+
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	return nil
+}
+
+func newDeployState(path string) *DeployState {
+	return &DeployState{
+		Path: path,
+		ASGs: []ASG{},
+	}
+}
+
+func (state *DeployState) gatherASGInfo(asgSvc *autoscaling.AutoScaling, eksAsgNames []string) error {
```

Comment:
NIT: Add function comments (this applies to all the action functions in this file)

Reply:
Fixed